src/components/InputUpload/InputUpload.tsx

```diff
@@ -0,0 +1,109 @@
+"use client";
+
+import React, { useRef, useState } from "react";
+import {
+  FieldValues,
+  Path,
+  UseFormReturn,
+  UseFormRegisterReturn,
+  PathValue,
+} from "react-hook-form";
+import Image from "next/image";
+import { IMAGE_TYPES } from "./InputUpload.constants";
+import * as S from "./InputUpload.styles";
+interface InputUploadProps<T extends FieldValues> {
+  position?: "relative" | "absolute";
+  isPreview?: boolean;
+  register: UseFormRegisterReturn;
+  setValue: UseFormReturn<T>["setValue"];
+  name: Path<T>;
+  children?: React.ReactNode;
+  onPreview?: (preview: string | null, imageSize: { width: number; height: number }) => void;
+}
+
+const InputUpload = <T extends FieldValues>({
+  position = "relative",
+  isPreview = true,
+  register,
+  setValue,
+  name,
+  children,
+  onPreview,
+}: InputUploadProps<T>) => {
+  const inputRef = useRef<null | HTMLInputElement>(null);
+  const [preview, setPreview] = useState<string | null>(null);
+  const [imageSize, setImageSize] = useState({
+    width: 0,
+    height: 0,
+  });
+
+  const handleFileChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    if (event.target.files) {
+      const file = event.target.files[0];
+      const fileType = file.type;
+
+      if (IMAGE_TYPES[fileType]) {
+        const reader = new FileReader();
+        reader.onloadend = () => {
+          const image = document.createElement("img");
+          image.onload = () => {
+            setImageSize({
+              width: image.width,
+              height: image.height,
+            });
+
+            if (onPreview) {
+              onPreview(reader.result as string, { width: image.width, height: image.height });
+            } else {
+              setPreview(reader.result as string);
+            }

```

Comment:
[ P2 ] 

코드가 증가 하더라도 as 를 없애는게 더 좋을것 같아서 한번 의견드려봐요!

```suggestion
        if (typeof reader.result !== "string") {
          return;
        }

        if (onPreview) {
          onPreview(reader.result, { width: image.width, height: image.height });
        } else {
          setPreview(reader.result);
        }
```

Reply:
얼리 리턴 한 번으로 as 문을 줄일 수 있는 건 너무 좋네요 ㅎㅎ (더 깊게 생각하지 못한 점 반성하게 되네요.. ㅠ) 😂